src/basic/build-path.c

```diff
@@ -45,7 +45,13 @@ static int get_runpath_from_dynamic(const ElfW(Dyn) *d, const char **ret) {
         if (!strtab)
                 return -ENOTRECOVERABLE;
 
-        /* According to dl.so runpath wins of both runpath and rpath are defined. */
+        /* On MIPS and RISC-V DT_STRTAB records an offset, not a valid address, so it has to be adjusted
+         * using the bias calculated earlier. */
+#if defined(__mips__) || defined(__riscv)
+        strtab += bias;
```

Comment:
please move that into the switch statement above. 

in C it's undefined behaviour to have a pointer to invalid memory, even if you never dereference it (with some additional rules). hence I think we should avoid having strtab point to some invalid memory initially, only to be corrected down here.

i.e. change the strab initialization above to:

```c
strtab = (const char *) ((size_t) d->d_un.d_val
#if defined(__mips__) || defined(__riscv)
                + bias
#endif
                );
```

or something like that.

i.e. do the addition while we are still dealing with numbers, and only convert to a pointer after we are done.

Reply:
the problem is that it will invalidate the "!strtab check", that's why I put this below it

Reply:
made the assignment itself conditional on !null